EnrollmentTrackUserPartition to be used for differentiated content

cahrens opened the pull request:
## [TNL-6674](https://openedx.atlassian.net/browse/TNL-6674)

### Description

Introduce EnrollmentTrackUserPartition for all courses.

### Sandbox
- [x] https://cahrens.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/d8a6192ade314473a78242dfeedfbf5b/4ed9eb74deee4d418d9047db0a326353/

On that unit, you can preview as "verified" to see content for verified learners, and "audit" to see content for audit learners. Note that this sandbox includes a hacky UX change (https://github.com/edx/edx-platform/pull/14731/commits/6eaef6f82aef3b2fdb3c28f0354985a8ea04f677) to allow setting of visibility of components by enrollment track. The complete implementation of that will be done in a future story, and the temporary commit is not part of this PR.

### Testing
- [x] Unit, integration, acceptance tests as appropriate
- [x] Analytics, no impact
- [ ] Performance, hopefully no impact! Not sure how to check this off though.

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @nasthagiri, whole PR
- [x] Code review: @staubina, second code reviewer

FYI: @staubina @jlajoie

### Post-review
- [x] Rebase and squash commits

cahrens commented:
jenkins run js

cahrens commented:
jenkins run js

cahrens commented on cms/djangoapps/contentstore/tests/test_utils.py:
```diff
@@ -493,12 +493,12 @@ def test_retrieves_partition_info_with_selected_groups(self):
             }
         ]
-        self.assertEqual(self._get_partition_info(), expected)
+        self.assertEqual(self._get_partition_info(schemes=["cohort", "random"]), expected)
```
Note to reviewers: exclude the enrollment_track partition.

cahrens commented on cms/djangoapps/contentstore/tests/utils.py:
```diff
@@ -299,8 +299,14 @@ def assertCoursesEqual(self, course1_id, course2_id):
-            course2_item = self.store.get_item(course2_item_loc)
-
+                course2_item = self.store.get_item(course2_item_loc)
+                # The EnrollmentTrackuserPartition stores the course_id, which will vary split vs. old mongo.
```
Note to reviewers: cloning an old mongo course into the new modulestore caused me some grief because the course_id at the beginning of the operation is the old ID, and then it changes at some point to the new (split) ID. I updated the course_module code to handle that, and then unit tests calling this method started to fail because the "old" course was not equal to the "new" course.

cahrens commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -296,7 +296,8 @@ def get_group_display_name(user_partitions, xblock_display_name):
     for user_partition in user_partitions:
         for group in user_partition['groups']:
-            if str(group['id']) in xblock_display_name:
+            # Temporary hack until TNL-6731 is addressed.
```
Note to reviewers: this issue will be fixed before this PR merges. FYI @mushtaqak

mushtaqak replied:
#14748 is merged :)

cahrens commented on cms/djangoapps/contentstore/views/tests/test_item.py:
```diff
@@ -343,13 +343,13 @@ def test_invalid_paging(self, page_number, page_size):
         self.course.user_partitions = [
             UserPartition(
-                id=0,
+                id=100,
```
Note to reviewers: Changing IDs so they more accurately represent what the code does (and avoid non-unique IDs).

cahrens commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -1775,7 +1775,7 @@ def test_certificate_generation_for_students(self):
             'skipped': 2
         }
-        with self.assertNumQueries(168):
+        with self.assertNumQueries(184):
```
Note to reviewers: there are 8 different learners (increase of 2 queries per learner).

nasthagiri replied:
Side note: Yikes.  No wonder certificate generation is slow.  We need to optimize this via bulk queries.

cahrens commented on common/djangoapps/course_modes/admin.py:
```diff
@@ -216,5 +227,24 @@ class CourseModeExpirationConfigAdmin(admin.ModelAdmin):
         model = CourseModeExpirationConfig
 
+
+class EnrollmentModeAdmin(admin.ModelAdmin):
```
Added the admin table just for easy viewing of the options (useful to make sure the migration ran correctly). Here it is (user staff has access): https://cahrens.sandbox.edx.org/admin/course_modes/enrollmentmode/

cahrens commented on common/djangoapps/course_modes/admin.py:
```diff
@@ -29,12 +29,15 @@
+    pass
 
+if not COURSE_MODE_SLUG_CHOICES:
+    # The entries in this table are populated in a migration file, and migration files do not
```
This is sad-- I was hoping to delete the hardcoded COURSE_MODE_SLUG_CHOICES, but can't because of unit tests.

cahrens commented on common/djangoapps/course_modes/models.py:
```diff
@@ -26,6 +26,25 @@
                   ])
 
 
+class EnrollmentMode(models.Model):
```
@bderusha what do you think of this model (and admin stuff above)? Do you like the name? Can you think of any way to link mode_slug in this model to mode_slug in CourseMode below?

cahrens replied:
This is gone, but I added a unit test to check that settings.COURSE_ENROLLMENT_MODES has integers under 100.

cahrens commented on lms/djangoapps/courseware/testutils.py:
```diff
@@ -197,6 +199,13 @@ def test_success_enrolled_staff(self, default_store, mongo_calls):
             with check_mongo_calls(mongo_calls):
                 self.verify_response()
 
+    def get_success_enrolled_staff_mongo_count(self):
```
See lms/djangoapps/lti_provider/tests/test_views.py below. I didn't get all the way to the root cause here, but given that it is old mongo only (and lti only), I didn't think it was worth spending time on it.

cahrens replied:
I agree that would be nicer, but this is a mixin, so overriding doesn't work.

cahrens commented on openedx/core/djangoapps/credit/tests/test_verification_access.py:
```diff
@@ -80,8 +80,9 @@ def test_creates_user_partitions(self):
         # Check that a new user partition was created for the ICRV block
-        self.assertEqual(len(self.course.user_partitions), 1)
-        partition = self.course.user_partitions[0]
+        partitions = _get_non_enrollment_track_partitions(self.course)
```
Note to reviewers: the VerificationUserPartition will be going away soon, but in the meantime I have to keep tests passing!

cahrens commented on openedx/core/djangoapps/verified_track_content/admin.py:
```diff
@@ -4,8 +4,8 @@
 
-from verified_track_content.forms import VerifiedTrackCourseForm
-from verified_track_content.models import VerifiedTrackCohortedCourse
+from openedx.core.djangoapps.verified_track_content.forms import VerifiedTrackCourseForm
```
Move of verified_track_content from LMS to openedx because of CMS unit test dependencies.

nasthagiri commented on openedx/core/djangoapps/credit/tests/test_verification_access.py:
```diff
@@ -272,4 +276,9 @@ def test_can_write_on_publish_signal(self):
+
+
+def _get_non_enrollment_track_partitions(course):
+    """ Return the user_partitions that are not of type "enrollment_track". """
```
Nit: Multi-line docstring, no?

cahrens replied:
Backed out this change.

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,147 @@
+"""
+UserPartitionScheme for enrollment tracks.
+"""
+from courseware.masquerade import (  # pylint: disable=import-error
```
What is the `import-error` complaining about?

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,147 @@
+# EnrollmentMode IDs will be low integers (starting at 0), which is important so that they
+# do not overlap with Groups in the CohortUserPartition or the RandomUserPartitionScheme
+# (CMS' course_group_config uses a minimum value of 100 for all generated IDs).
+ENROLLMENT_GROUP_IDS = {}
```
Does this need to be a global constant, or can it be a function instead?  Input: enrollment_mode; Output: group_id?

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,147 @@
+except:  # pylint: disable=bare-except
+    pass
+
+if not ENROLLMENT_GROUP_IDS:
```
Once again, can we move this to `lms/env/test.py` instead, so it's not part of production code?

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,147 @@
+            # The user is masquerading as a generic student. We can't show any particular group.
+            return None
+
+        if is_course_using_cohort_instead(course_key):
```
Shouldn't this check be done first (before the masquerading logic)?

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,147 @@
+            return None
+        mode_slug, is_active = CourseEnrollment.enrollment_mode_for_user(user, course_key)
+        if mode_slug and is_active:
+            course_mode = CourseMode.mode_for_course(course_key, mode_slug)
```
This may not work for old courses (as [tagged on the bulk-email PR](https://github.com/edx/edx-platform/pull/14688/files#r105997547)).  The implementation of `mode_for_course` calls `modes_for_course` with the limiting default values for `include_expired` and `only_selectable`.

cahrens replied:
Yikes, good catch! I'll have to figure out what to do here.

cahrens replied:
OK, passing in all modes.

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,147 @@
+            group = Group(ENROLLMENT_GROUP_IDS[mode.slug], unicode(mode.name))
+            all_groups.append(group)
+
+        return all_groups
```
Nitty: Consider using a list comprehension syntax for readability, consolidating the above 6 lines of code.

```
return [
    Group(ENROLLMENT_GROUP_IDS[mode.slug], unicode(mode.name))
    for mode in CourseMode.modes_for_course(course_key, include_expired=True, only_selectable=False)
]
```

cahrens replied:
Yikes, that seems a lot less readable to me!

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,147 @@
+        'Serialize' to a json-serializable representation.
+
+        Returns:
+            a dictionary with keys for the properties of the partition, with an empty array for
```
Nitty: capitalize sentence.

cahrens commented on cms/djangoapps/contentstore/tests/utils.py:
```diff
@@ -299,8 +299,14 @@ def assertCoursesEqual(self, course1_id, course2_id):
+                # The EnrollmentTrackuserPartition stores the course_id, which will vary split vs. old mongo.
+                # Also at least one test calls this comparison method with a purposefully different course.
+                # Clear stored course_id so that own_metadata call below does not fail.
+                course1_item.user_partitions[0].parameters['course_id'] = None
```
The problem is that we need the course_id to return the groups from the user partition (and groups is a property). I don't see how we can pass the course_id unless I change groups completely in a method.

UserPartitions (after initial construction) are constructed via the from_json method. We don't have the course_id available then either.

nasthagiri replied:
Yes.  Exactly.  UserPartitions don't need to always know the `course_id` context.  They can be provided the context by the caller when needed.

cahrens replied:
course_id is no longer persisted. We don't call to_json on the partition anymore, but just to be clear, I'm also explicitly not persisting the course_id in the to_json method.

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -910,6 +918,48 @@ def __init__(self, *args, **kwargs):
+            enrollment_track_partitions = self.get_user_partitions_for_scheme(enrollment_track_scheme)
+            # Note that when a course_module is initially created, the ID is the draft version.
+            # We just want to store the course_id without published or draft.
+            course_id = unicode(self.id.for_branch(None))
```
As I mentioned above, both the scheme and the partition need access to the course_id. Let's talk in person about options.

cahrens replied:
FYI, I'm working on this refactor to change group from a property to a method on a separate branch, as it impacts a lot of files and I expect I won't get them all right the first time.

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -910,6 +918,48 @@ def __init__(self, *args, **kwargs):
+                used_ids = set(p.id for p in self.user_partitions)
+                while new_id in used_ids:
+                    new_id += 1
+                create_partition = True
```
Yes, if this is called by LMS it will update the course module. But the init method in general sets up a lot of default values, so doesn't the same situation occur for those? Let's chat.

cahrens replied:
We talked about this in person, but recording here too. I don't think there is a race condition issue because each course_module instance that is created will persist its own version of user_partitions. It is possible (and quite likely in production) that multiple ones will add a enrollment_track partition to their user_partitions variable, but they will all be equivalent enrollment_track partitions, and which is the last one saved will just overwrite all the others. There will be no merging of user_partitions variables across multiple instances of course_module.

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -910,6 +918,48 @@ def __init__(self, *args, **kwargs):
+                    )
+                )
+
+        except UserPartitionError:
```
Sure, will make it just around UserPartition.get_scheme. This will probably also change soon as we are going to make the code robust to missing schemes (as part of the ICRV removal).

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -909,6 +917,8 @@ def __init__(self, *args, **kwargs):
         if self.discussion_topics == {}:
             self.discussion_topics = {_('General'): {'id': self.location.html_id()}}
 
+        self.set_default_user_partition()
```
@cpennington @ormsbee What are your thoughts about persisting a change to course_module via the init method? @nasthagiri and I have been talking about it, and we are not sure if there is a better solution.

The background here is that we are introducing a new UserPartition type that is based on enrollment track (so we can ultimately get rid of the cohorting/enrollment track beta feature and instead use a UserPartition that dynamically gets the learner's enrollment mode). We want this partition to be available for all courses.

With the code as-is, if the course_module is loaded from LMS and no enrollment track partition exists, we will create one. This means that the "published" version of the course will end up persisting user_partitions through the save method in caching_descriptor_system's "xblock_from_json" method.

There are other fields already being set in this way (discussion_topics, grading policy, etc.), so I don't think this is a new issue. But is there a better way to do this?

cahrens commented on common/djangoapps/course_modes/tests/test_models.py:
```diff
@@ -471,3 +471,22 @@ def test_verified_min_price(self, mode_slug, is_error_expected):
+    """
+    def tearDown(self):
+        super(EnrollmentModeTest, self).tearDown()
+        EnrollmentMode.objects.all().delete()
```
@jzoldak is this true? There are a ton of unit tests that delete models in the tearDown. Are all of those unnecessary?

jzoldak replied:
@cahrens here's the best link I could find that explains [default Django testing behavior](https://docs.djangoproject.com/en/1.8/topics/testing/advanced/#using-different-testing-frameworks)

This behavior involves:
1. Performing global pre-test setup.
2. Looking for tests in any file below the current directory whose name matches the pattern test*.py.
3. Creating the test databases.
4. Running migrate to install models and initial data into the test databases.
5. Running the tests that were found.
6. Destroying the test databases.
7. Performing global post-test teardown.

ormsbee replied:
@cahrens: @nasthagiri's right. As long as you're inheriting from Django's `TestCase`, individual test methods will be run in a transaction and rolled back at the end, so:

```
start transaction
setUp()
test_method_1()
tearDown()
rollback

start transaction
setUp()
test_method_2()
tearDown()
rollback

etc.
```

cahrens replied:
Test class is gone.

cahrens commented on common/lib/xmodule/xmodule/partitions/partitions.py:
```diff
@@ -2,12 +2,23 @@
 
 
+class UserPartitionList(List):
+    """Special List class for listing UserPartitions"""
```
Sigh, you are correct. I just hate to take up the space (and am generally grumpy about having to write documentation lines for things that are self-documenting, as self-documenting should be the goal).

But I suppose I will change all my single-line documentation in this PR. :(

nasthagiri replied:
_> I just hate to take up the space .. for things that are self-documenting_

You don't have to!  If it's self-documenting, you are smarter than the linter.  Just pragma (pylint-disable) the warning!  This is approved by Clean Code standards - so you'll still be a "conformant rebel".  :)

cpennington replied:
Also, if either of you think that docstrings aren't worth linting on, drop a note here: https://github.com/edx/edx-lint/pull/53.

cahrens commented:
@nasthagiri I have the partition service approach working (on my devstack). There are a lot of tests to clean up, doc strings to write, and all, but can you please take an early look and let me know what you think? Because there are so many files in this PR, I'll tag you on each of the files that you actually need to look at.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -84,7 +85,7 @@ def assign_group_ids(self):
         Assign ids for the group_configuration's groups.
         """
-        used_ids = [g.id for p in self.course.user_partitions for g in p.groups]
+        used_ids = [g.id for p in get_course_user_partitions(self.course) for g in p.groups]
```
@nasthagiri I chose to not replace all usages of self.course.user_partitions, as most calls really only want the partitions stored in course.user_partitions (for example, when looking specifically at cohort, random, or verification partitions).

When I did need all partitions, I used this helper method get_course_user_partitions unless I didn't have access to the course, in which case I went through the runtime service.

This particular call needs all partitions because it is what the Studio component visibility dialog uses (though this file is actually one that would be removed before merging the PR, as it is only present for local testing).

nasthagiri replied:
Nit: I wonder if there's a better name for `get_course_user_partitions`?  It was unclear from the name whether this returned only the partitions stored in the course content, or also included the dynamic partitions on the course?
Maybe rename to `get_all_partitions_for_course`?

cahrens replied:
Yes, I agree. I'll change it to get_all_partitions_for_course.

cahrens commented on cms/djangoapps/contentstore/views/preview.py:
```diff
@@ -213,10 +214,25 @@ def _preview_module_system(request, descriptor, field_data):
             "i18n": ModuleI18nService,
             "settings": SettingsService(),
             "user": DjangoXBlockUserService(request.user),
+            "partitions": PreviewPartitionService(user=request.user, course_id=course_id)
```
@nasthagiri I needed to add a Studio version of the service or else viewing the components in Studio failed (due to mixin further down).

cahrens commented on cms/djangoapps/contentstore/views/preview.py:
```diff
@@ -213,10 +214,25 @@ def _preview_module_system(request, descriptor, field_data):
 
+class PreviewPartitionService(PartitionService):
+    """
+    Another runtime mixin that provides access to the student partitions defined on the
```
@nasthagiri These comments will be updated

Note that the implementation is now identical to to the LMSPartitionService. But it's so little code, that I don't think it matters.

cahrens commented on lms/djangoapps/course_blocks/transformers/user_partitions.py:
```diff
@@ -48,7 +49,7 @@ def collect(cls, block_structure):
         user_partitions = [
             user_partition
-            for user_partition in getattr(root_block, 'user_partitions', [])
+            for user_partition in get_course_user_partitions(root_block)
```
@nasthagiri transformer code change. I am assuming "root_block" here is the course.

cahrens commented on lms/djangoapps/lms_xblock/mixin.py:
```diff
@@ -26,6 +27,7 @@ def to_json(self, access_dict):
             return {unicode(k): access_dict[k] for k in access_dict}
 
 
+@XBlock.wants('partitions')
```
@nasthagiri This is the mixin, which needs to reference the runtime service (no direct reference to course here).

cahrens commented on openedx/core/lib/partitions/partitions_service.py:
```diff
@@ -3,7 +3,57 @@
+ENROLLMENT_TRACK_PARTITION_ID = 50
+
+
+def get_course_user_partitions(course):
```
@nasthagiri here is the helper method, get_course_user_partitions.

cahrens commented on openedx/core/lib/partitions/partitions_service.py:
```diff
@@ -3,7 +3,57 @@
+        return
+
+    used_ids = set(p.id for p in course.user_partitions)
+    if ENROLLMENT_TRACK_PARTITION_ID in used_ids:
```
@nasthagiri the ID situation here is tricky. I can't just pick an unused ID anymore because this partition/ID is never persisted, and xblock's group_access WILL store the partition ID, so it is crucial that the ID always be the same whenever this dynamic partition is returned.

Luckily we should have 0-99 free as indices, so I think that dynamic partitions just have to declare their IDs as constants within that range (and if an XML-authored course happens to hit one of these IDs, then we have to skip adding those dynamic partitions). Of course, this would require much commenting, but for now I'm using "50".

cahrens commented on lms/djangoapps/lms_xblock/mixin.py:
```diff
@@ -131,7 +133,7 @@ def _get_user_partition(self, user_partition_id):
         `NoSuchUserPartitionError` if the lookup fails.
         """
-        for user_partition in self.user_partitions:
+        for user_partition in self.runtime.service(self, 'partitions').course_partitions:
```
@nasthagiri this method is what the access control code calls. Previously, "self" was any xblock, not just the course. I am now looking at on the course user_partitions, but I think that is fine since that is the only place they should be modified. 

For reference, here is the access usage: https://github.com/edx/edx-platform/blob/b287f48f8b9ea31e2ad42be8f609dbb0e0059ef9/lms/djangoapps/courseware/access.py#L496

cahrens commented on cms/djangoapps/contentstore/views/preview.py:
```diff
@@ -213,10 +214,27 @@ def _preview_module_system(request, descriptor, field_data):
+    def get_course(self):
+        return modulestore().get_course(self._course_id)
+
+    def get_user_group_id_for_partition(self, user_partition_id):
```
We have to override this method because now when running in Studio a partitions runtime service actually exists. Previously, this code would just return None when trying to get the service (and the code was robust to that):

https://github.com/edx/edx-platform/blob/eea25184521c4fe224adc676006f6d86edb8e215/common/lib/xmodule/xmodule/split_test_module.py#L195

cahrens replied:
Rename as StudioPartitionService.

cahrens commented on lms/djangoapps/courseware/tests/test_group_access.py:
```diff
@@ -183,8 +184,13 @@ def check_access(self, user, block_location, is_accessible):
         """
+        # Set the partition service on the item.
+        item = modulestore().get_item(block_location)
+        item.runtime._services['partitions'] = LmsPartitionService(  # pylint: disable=protected-access
```
@nasthagiri and @cpennington I'm struggling with how best to set the partition service for tests that depend on checking group access. There seem to be a variety of ways that tests set services, and I don't understand the differences.

In this particular case, you can see I just stuck it direction on item.runtime._services. That seems pretty ugly.

I'll add some comments to other ways this is done, as I'd like to make sure I'm not introducing technical debt into the tests.

cahrens replied:
@cpennington @nasthagiri just hit another partition service unit test problem that I don't know how to solve. It's related to discussion component visibility, and I can't bind a partition service to an actual xblock instance because the unit tests are testing APIs that read all the discussion xblocks from the course module and then iterate through them.

cahrens commented on common/lib/xmodule/xmodule/tests/test_split_test_module.py:
```diff
@@ -80,29 +82,35 @@ def setUp(self):
+                MockUserPartitionScheme()
+            )
+        ]
+        partitions_service = MockPartitionService(
```
@nasthagiri @cpennington Here is another example-- a PartitionService is being added onto module_system._services, and then module_system is bound to the split_test_module instance and user. module_system comes from get_test_system.

cahrens commented on lms/djangoapps/courseware/tests/test_access.py:
```diff
@@ -314,7 +315,12 @@ def test_has_access_in_preview_mode_with_group(self):
         chapter.group_access = {partition_id: [group_0_id]}
-        chapter.user_partitions = self.course.user_partitions
+
+        module_system = get_test_system()
```
@cpennington @nasthagiri Same sort of pattern as the last test case, though here I'm using LmsPartitionService becuase the test lives in LMS.

cahrens commented on common/djangoapps/course_modes/admin.py:
```diff
@@ -29,12 +29,15 @@
 
+if not COURSE_MODE_SLUG_CHOICES:
+    # The entries in this table are populated in a migration file, and migration files do not
+    # run as part of unit tests. Therefore, for tests hard-code the choices.
```
I was chatting with @ormsbee, and there are general problems with me trying to access a database at the time that the file is loaded (with future versions of django). Plus edx-platform would have to be restarted for any new EnrollmentMode instances to be recognized (both here and in verified_track_content/partition_scheme.py).

I could try to make both of these usages lazy, but then I still have to figure out how to populate them in all unit tests which need them (a much bigger deal for verified_track_content/partition_scheme.py than for the admin table!).

@bderusha sorry, but I think I am going to abandon the database table approach and go with a setting that has the slug/ID mapping. :(

cahrens commented:
@cpennington @nasthagiri Should LmsPartitionService also be registered in the LmsModuleSystem instance created within module_render?

https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/module_render.py#L680

I don't know when this is used vs. the LmsModuleSystem defined in lms/djangoapps/lms_xblock/runtime.py (which is the currently the only place in LMS where LmsPartitionService is registered).

cahrens commented on common/djangoapps/course_modes/admin.py:
```diff
@@ -216,5 +227,24 @@ class CourseModeExpirationConfigAdmin(admin.ModelAdmin):
+    class Meta(object):
+        model = EnrollmentMode
+
+    def has_add_permission(self, request):
```
I deleted the EnrollmentMode model and associated admin view.

cahrens commented on common/djangoapps/course_modes/migrations/0008_enrollmentmode.py:
```diff
@@ -0,0 +1,46 @@
+            objects.create(mode_slug=mode)
+
+
+def delete_modes(apps, schema_editor):
```
Migration is gone.

cahrens commented on lms/djangoapps/lms_xblock/mixin.py:
```diff
@@ -131,7 +133,8 @@ def _get_user_partition(self, user_partition_id):
         `NoSuchUserPartitionError` if the lookup fails.
         """
-        for user_partition in self.user_partitions:
+        partition_service = self.runtime.service(self, 'partitions')
```
Defaulting to self.user_partitions should fix all my failing unit tests, but I don't feel great about it. Though if I don't provide this fallback, I should change the decorator to _XBlock.needs_.

@cpennington @nasthagiri thoughts?

cahrens commented:
FYI @cpennington, @nasthagiri and I chatted and came up with an approach to ensure that a PartitionService is always available, even for xblocks obtained directly from the modulestores (for unit tests). It's going to mean moving stuff around and not requiring a user on a PartitionService, but it all seems clear.

Just wanted to let you know since I had tagged you about testing concerns.

cahrens commented:
jenkins run python

nasthagiri commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -3,41 +3,99 @@
+
+    used_ids = set(p.id for p in course.user_partitions)
+    if ENROLLMENT_TRACK_PARTITION_ID in used_ids:
+        log.warning("Cannot add 'enrollment_track' partition, as ID {id} is assigned to {partition}".format(
```
Can you include the course_id in the log statement so it's available in Splunk?

cahrens replied:
Good idea, will do.

nasthagiri commented:
Agreed.  This looks much cleaner now.  Great work.

cahrens commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -3,41 +3,99 @@
         If not, assign them to one of the groups, persist that decision, and
         return the group_id.
 
-        If the group they are assigned to doesn't exist anymore, re-assign to one of
```
Note to reviewers: I don't see that the code actually does this. Perhaps the random user partition scheme does, but that is implementation-specific to a specific scheme.

cahrens commented on lms/djangoapps/courseware/access.py:
```diff
@@ -466,12 +466,6 @@ def _has_group_access(descriptor, user, course_key):
     This function returns a boolean indicating whether or not `user` has
     sufficient group memberships to "load" a block (the `descriptor`)
     """
-    if len(descriptor.user_partitions) == len(get_split_user_partitions(descriptor.user_partitions)):
```
Note to reviewers: we don't want this short-circuit anymore, as dynamic user partitions may be present.

nasthagiri replied:
Hopefully this doesn't cause a noticeable performance issue.  We *really* need to transition over to using (the optimized) transformers soon for all course access control.

cahrens commented on lms/djangoapps/courseware/tests/test_access.py:
```diff
@@ -431,6 +430,7 @@ def test__has_access_descriptor(self):
         user = Mock()
         descriptor = Mock(user_partitions=[])
         descriptor._class_tags = {}
+        descriptor.merged_group_access = {}
```
Note to reviewers: needed because of the removal of the short circuit logic in access.py.

cahrens commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -295,9 +296,10 @@ def get_group_display_name(user_partitions, xblock_display_name):
-        for group in user_partition['groups']:
-            if str(group['id']) in xblock_display_name:
-                return group['name']
+        if user_partition['scheme'] == 'random':
```
Note to reviewers: this code change is going in soon via #14748.

mushtaqak replied:
#14748 is merged :)

cahrens commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -408,7 +410,7 @@ def get_user_partition_info(xblock, schemes=None, course=None):
             partitions.append({
                 "id": p.id,
-                "name": p.name,
+                "name": unicode(p.name),  # Convert into a string in case ugettext_lazy was used
```
Note to reviewers: I had to add this because the EnrollmentTrackUserPartition uses ugettext_lazy (due to the fact that it lives in common.lib.xmodule.xmodule). When CMS unit tests ran, p.name was a function instead of a string, and JSON encoding failed.

cahrens commented on cms/envs/bok_choy.py:
```diff
@@ -93,6 +93,8 @@
 FEATURES['ENABLE_MOBILE_REST_API'] = True  # Enable video bumper in Studio
 FEATURES['ENABLE_VIDEO_BUMPER'] = True  # Enable video bumper in Studio settings
 
+FEATURES['ENABLE_ENROLLMENT_TRACK_USER_PARTITION'] = True
```
Note to reviewers: feature flag was introduced because the next story will add the ability to set component visibility based on this new partition. @sstack22 doesn't want course authors seeing this functionality before the other parts (including the ability to segment discussions) are present.

Hopefully this feature flag will be deleted once everything is complete, as our plan is to make the feature invisible for courses where only a single enrollment track exists.

nasthagiri replied:
By the way, in the future, consider using a waffle switch in these cases.  This way, you wouldn't need to be blocked on a release in order to enable/disable the feature.  I've been using it for several features now, and created helper waffle utility classes: https://github.com/edx/edx-platform/blob/master/openedx/core/djangolib/waffle_utils.py

cahrens replied:
Thanks for the pointer-- definitely next time!

cahrens commented:
jenkins run js

cahrens commented:
jenkins run python

cahrens commented:
jenkins run js

cahrens commented:
jenkins run a11y

cahrens commented:
Trying to get Jenkins happy again.

cahrens commented:
jenkins run a11y

cahrens commented:
jenkins run js

cahrens commented:
jenkins run bokchoy

cahrens commented:
jenkins run lettuce

cahrens commented:
@staubina and @nasthagiri this PR is ready for re-review, and the sandbox has been updated (the link in the Description section is up-to-date).

cahrens commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -3,41 +3,114 @@
+
+
+# settings will not be available when running nosetests.
+FEATURES = getattr(settings, 'FEATURES', {})
```
This is the same thing that was recently introduced into capa, so hopefully it is an acceptable pattern.

https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/capa_base.py#L45

cahrens commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -3,41 +3,114 @@
 persist the assignments.
 """
-from abc import ABCMeta, abstractproperty
+from django.conf import settings
```
@nasthagiri are these django imports OK in this file? Other files in common/lib/module use this workaround instead of importing django.utils.translation:

````
# Make '_' a no-op so we can scrape strings. Using lambda instead of
#  `django.utils.translation.ugettext_noop` because Django cannot be imported in this file
_ = lambda text: text

nasthagiri replied:
Yes, it's Ok.  Just as modulestore/django.py is django-dependent, we can say that the partitions folder is also django-dependent.

cahrens commented:
jenkins run bokchoy

cahrens commented:
jenkins run bokchoy

staubina commented on common/lib/xmodule/xmodule/partitions/tests/test_partitions.py:
```diff
@@ -509,10 +530,59 @@ def test_get_group(self):
+    def setUp(self):
+        super(TestGetCourseUserPartitions, self).setUp()
+        # django.conf.settings is not available when nosetests are run
+        FEATURES['ENABLE_ENROLLMENT_TRACK_USER_PARTITION'] = True
```
nit: Define 'ENABLE_ENROLLMENT_TRACK_USER_PARTITION' as a static because it is used more than once.

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -17,6 +17,7 @@
 from xmodule.course_metadata_utils import DEFAULT_START_DATE
 from xmodule.graders import grader_from_conf
 from xmodule.mixin import LicenseMixin
+from xmodule.partitions.partitions import UserPartition
```
Not really-- it's referenced in doc, but I think the import can be removed.

nasthagiri replied:
Reminder to remove this import.

cahrens commented on common/lib/xmodule/xmodule/modulestore/inheritance.py:
```diff
@@ -5,10 +5,10 @@
 from xblock.fields import Scope, Boolean, String, Float, XBlockMixin, Dict, Integer, List
 from xblock.runtime import KeyValueStore, KvsFieldData
 from xmodule.fields import Date, Timedelta
+from xmodule.partitions.partitions import UserPartition
```
I can move this import back to its original location so this file isn't part of the change set.

nasthagiri replied:
No worries.  Small change.

staubina commented:
Just some minor comments, nothing major from what I can see. 

Tested on the Sandbox also and it looks good.

nasthagiri commented on cms/djangoapps/contentstore/views/tests/test_item.py:
```diff
@@ -343,13 +345,13 @@ def test_invalid_paging(self, page_number, page_size):
         self.course.user_partitions = [
             UserPartition(
-                id=0,
+                id=100,
```
Can you add a comment to explain why this number?

cahrens replied:
It actually could stay at 0 now, since the enrollment track partition is at 50. I was just trying to make the numbers more realistic, as they would have a minimum value of 100.

I don't want to put comments throughout the whole file, but I will put a place to put a general comment about IDs and the different ranges of numbers.

nasthagiri commented on cms/envs/common.py:
```diff
@@ -225,6 +227,9 @@
     'ALLOW_PUBLIC_ACCOUNT_CREATION': True,
+
+    # Whether or not the dynamic EnrollmentTrackUserPartition should be registered.
+    'ENABLE_ENROLLMENT_TRACK_USER_PARTITION': False
```
Nit: Add ending commas here and above.

cahrens commented:
@nasthagiri and @staubina I have addressed the review feedback and rebased on master.

staubina commented:
👍

nasthagiri commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -3,41 +3,114 @@
+        log.warning("No 'enrollment_track' scheme registered, EnrollmentTrackUserPartition will not be created.")
+        return None
+
+    used_ids = set(p.id for p in course.user_partitions)
```
Do we also prevent new partitions from being created with the same ID as `ENROLLMENT_TRACK_PARTITION_ID`?  I'm concerned of long downstream issues where we have this check in place every time `get_all_partitions_for_course ` is called, and we silently fail, allowing a course-team/developer to override an already existing partition.  This can result in potentially hard-to-debug and unexpected conditions - where a new partition overtakes the enrollment-track partition and starts using previously configured user-access mappings.
Why not raise an exception here instead?
If you're concerned about initial rollout and worried there may be courses already using this partition, then, you can:
1. A priori: Run a mongo query to find those courses beforehand.
2. Post priori: Release this initially with a silent failure, but monitor splunk.  And then change this to an exception after rollout (with a TNL ticket reminding to change this).  However, you have no guarantee that all courses have been accessed by that time.
3. Put the raise exception behind a waffle flag so you can disable it quickly if needed.

cahrens replied:
@nasthagiri I have thought more about this and am coming around to an exception. We are guaranteed that partitions created in Studio won't conflict (they are > 100), but partitions created in some other way could potentially conflict.

I like option #2. I will create a ticket for it in our epic.

Note that when we get to the discussions work we will likely want to throw an exception if 2 user partitions exist with a group that has the same ID. That's another place where we should be fine for partitions created by Studio, but could in the future run into difficulty.

nasthagiri commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -77,18 +150,28 @@ def _get_user_partition(self, user_partition_id):
+        if partition.id == user_partition_id:
+            return partition
+
+    return None
```
By the way, python always returns `None` by default.  So this line unnecessary.

cahrens replied:
I know, it's just clearer to me.

nasthagiri commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -3,41 +3,114 @@
+FEATURES = getattr(settings, 'FEATURES', {})
+
+
+def get_all_partitions_for_course(course):
```
Do you expect all callers to remember to check for the `.is_active` field on each partition?  Or, do you want to add 'check_for_active` as an optional parameter here?

cahrens replied:
That's a good idea-- it will simplify some of our other code and remind people that "active" exists (something which I myself forgot). Will do.

nasthagiri commented on lms/envs/common.py:
```diff
@@ -3068,3 +3071,13 @@
+    "professional": 3,
+    "no-id-professional": 4,
+    "credit": 5,
+    "honor": 6
```
Nitty: ending comma

nasthagiri commented on lms/envs/common.py:
```diff
@@ -2146,8 +2149,8 @@
 
-    # Verified Track Content Cohorting
-    'verified_track_content',
+    # Verified Track Content Cohorting (Beta feature that will hopefully be removed)
```
Is there a TNL ticket for this cleanup task?  Consider adding the ticket number to this comment.

cahrens replied:
No, there isn't one yet. I think creating it is a bit premature, as we have to see how the migration of courses (at rerun time) ends up going.

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,138 @@
+            course_mode = CourseMode.mode_for_course(
+                course_key,
+                mode_slug,
+                modes=CourseMode.modes_for_course(course_key, include_expired=True, only_selectable=False)
```
Nitty: ending comma

nasthagiri commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -0,0 +1,138 @@
+                modes=CourseMode.modes_for_course(course_key, include_expired=True, only_selectable=False)
+            )
+            if not course_mode:
+                course_mode = CourseMode.DEFAULT_MODE
```
Shouldn't the logic of returning the default mode be included within `CourseMode.mode_for_course`?

cahrens replied:
It's not-- that method returns None if the mode requested does not exist.

nasthagiri commented on lms/djangoapps/courseware/tests/test_access.py:
```diff
@@ -301,9 +301,9 @@ def test_has_access_in_preview_mode_with_group(self):
-        partition_id = 0
-        group_0_id = 0
-        group_1_id = 1
+        partition_id = 100
```
Here's another case of using the magic 100 value.  Consider adding a constant for this somewhere so we can use it in all affected tests - and the comment to explain why would be in that single place where the constant is defined.

cahrens commented:
@nasthagiri @staubina I have addressed the most recent feedback and rebased on master. If you are interested in looking at the changes, see the last 3 commits.

I plan to squash commits once I get a clean build on this, so probably around 12:30 PM.

cahrens commented:
jenkins run js

cpennington commented:
*EdX Release Notice*: This PR has been deployed to production. (We had an infrastructure issue, so this message was posted manually.)
